web/pdf_page_view.js

```diff
@@ -1067,74 +991,68 @@ class PDFPageView {
     const transform = outputScale.scaled
       ? [outputScale.sx, 0, 0, outputScale.sy, 0, 0]
       : null;
-    const renderContext = {
-      canvasContext: ctx,
-      transform,
-      viewport,
-      annotationMode: this.#annotationMode,
-      optionalContentConfigPromise: this._optionalContentConfigPromise,
-      annotationCanvasMap: this._annotationCanvasMap,
-      pageColors,
-      isEditing: this.#isEditing,
-    };
-    const renderTask = (this.renderTask = pdfPage.render(renderContext));
-    renderTask.onContinue = renderContinueCallback;
-
-    const resultPromise = renderTask.promise.then(
-      async () => {
-        showCanvas?.(true);
-        await this.#finishRenderTask(renderTask);
-
-        this.structTreeLayer ||= new StructTreeLayerBuilder(
-          pdfPage,
-          viewport.rawDims
-        );
-
-        this.#renderTextLayer();
-
-        if (this.annotationLayer) {
-          await this.#renderAnnotationLayer();
+    const resultPromise = this._drawCanvas(
+      this._getRenderingContext(ctx, transform),
+      () => {
+        prevCanvas?.remove();
+        this._resetCanvas();
+      },
+      renderTask => {
+        // Ensure that the thumbnails won't become partially (or fully) blank,
+        // for documents that contain interactive form elements.
+        this.#useThumbnailCanvas.regularAnnotations =
+          !renderTask.separateAnnots;
+
+        // If there is a `.detailView` that still needs to be rendered, it will
+        // dispatch the pagerendered event once it's done.
+        if (
+          !this.detailView ||
+          this.detailView.renderingState === RenderingStates.FINISHED
+        ) {
+          this.dispatchPageRendered(
+            /* cssTransform */ false,
+            /* isDetailView */ false
+          );
         }
```

Comment:
When there's a detailView, won't this condition effectively prevent a "pagerendered" event from ever being dispatched with `isDetailView = false` set?

Is that behaviour correct, since it means that we may no longer have a way to detect the initial *full* page rendering and the following check could theoretically always hold (and thus leave `_onePageRenderedCapability pending)?
https://github.com/mozilla/pdf.js/pull/19128/files#diff-4238fc7e96d72f4095859850f3d24a9298ec2dc8bd96437214a02dccbd48385cR876

Reply:
Mh true. I think when I originally wrote this PR I was trying to dispatch _one_ `pagerendered` event per page (whether it's an initial render or a re-render). With the `isDetailView` property on events it's indeed easier to just dispatch both times, and consumers can choose which ones to ignore.